### PR TITLE
BUG: str(np.float) should print with the same number of digits as python str(float)

### PR DESCRIPTION
**Pull request by ahaldane**
See https://github.com/pandas-dev/pandas/issues/18123

This makes numpy floating-point scalars `str` and `repr`  (ie, `str(np.float64(1.2345))`) print in the same way as python floats (done in the 2nd commit). That is:
 * ~~`str` uses 8 digits, in fixed mode for `1e-4 < x < 1e11` and scientific mode otherwise.~~
 * `repr` uses as many digits as needed to specify the value uniquely, in fixed mode for `1e-4 < x < 1e16` and scientific mode otherwise.

(Updated: Both `str` and `repr` now print all digits, like floats in python3.)

The PR is so big because I needed to expose control of the digit cutoff mode in the Dragon4 algorithm, which I had left hidden before. That's all in the 1st commit. This allows you to control whether `precision` refers to the number of digits past 0, or the number of significant digits. An example:
```python
>>> np.format_float_positional(0.00123456789, precision=4, fractional=True)
'0.0012'
>>> np.format_float_positional(0.00123456789, precision=4, fractional=False)
'0.001235'
```
This was required to fix the scalar `str` because Python outputs floats with `fractional=False`, while in numpy arrays we use `fractional=True`.

### CODE REVIEW

**Comment by ahaldane** [comment]
Interesting: In python3, the `str` of floats was changed to be identical to the `repr`, so it has enough digits to be uniquely specified.

See https://bugs.python.org/issue9337
https://stackoverflow.com/questions/25898733/why-does-strfloat-return-more-digits-in-python-3-than-python-2

So I guess the design decision here is, should we ouput the shorter `str` for numpy-floats in python2, or should we always use the exact floats? I'm going to say "always use the unique (longer) value".

**Comment by charris** [comment]
> "always use the unique (longer) value"

Sounds right to me. Python 2 is starting to recede in the rear mirror, time to prepare folks for 3 :0

File: numpy/core/src/multiarray/dragon4.c [hunk 1]
```diff
@@ -35,8 +35,12 @@
 #include <string.h>
 
 #include <assert.h>
-/* #define DEBUG_ASSERT(stmnt) assert(stmnt) */
+
+#if 0
+#define DEBUG_ASSERT(stmnt) assert(stmnt)
+#else
 #define DEBUG_ASSERT(stmnt) {}
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/dragon4.c]: This doesn't look safe to me. Shouldn't it be `do {} while(0)`?

File: numpy/core/src/multiarray/multiarraymodule.c [hunk 2]
```diff
@@ -3646,55 +3629,40 @@ dragon4_scientific(PyObject *NPY_UNUSED(dummy), PyObject *args, PyObject *kwds)
...
+    CutoffMode cutoff_mode;
+    DigitMode digit_mode;
     TrimMode trim = TrimMode_None;
-    int sign=0, unique=1;
+    int sign=0, unique=1, fractional=0;
 
-    if (!PyArg_ParseTupleAndKeywords(args, kwds, "O|iiisii", kwlist,
-                &obj, &precision, &unique, &sign, &trimstr, &pad_left,
-                &pad_right)) {
+    if (!PyArg_ParseTupleAndKeywords(args, kwds, "O|iiiisii", kwlist,
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/multiarraymodule.c]: This should have `:dragon4_position` after the list of types, to produce a more useful error.

File: numpy/core/src/multiarray/dragon4.c [hunk 3]
```diff
@@ -1187,8 +1170,7 @@ Dragon4(const npy_uint64 mantissa, const npy_int32 exponent,
      * increases the number. This will either correct digitExponent to an
      * accurate value or it will clamp it above the accurate value.
      */
-    if ( (cutoffMode == CutoffMode_FractionLength ||
-          cutoffMode == CutoffMode_MaxFractionUnique) &&
+    if (cutoffNumber >= 0 && cutoffMode == CutoffMode_FractionLength &&
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/dragon4.c]: Can you add a comment explaining the purpose of `cutoffNumber >= 0`? the only comment i can see in this diff about `cutoffNumber == -1` is a deletion!

**Comment by ahaldane** [comment]
Updated.

Although, curiously I see no effect of adding `:funcname` to the end of the thy `PyArg...` argument string.

**Comment by charris** [comment]
Is this ready?

**Comment by ahaldane** [comment]
As far as I am concerned, yes.

**Comment by charris** [comment]
OK, in it goes. Thanks Allan.

**Pull request merged**